crates/iceberg/src/spec/arrow.rs

```diff
@@ -0,0 +1,553 @@
+        let result = visit_type(field.data_type(), visitor)?;
+        visitor.after_field(field)?;
+        results.push(result);
+    }
+    visitor.schema(schema, results)
+}
+
+const ARROW_FIELD_ID_KEYS: [&str; 2] = ["PARQUET:field_id", "field_id"];
+const ARROW_FIELD_DOC_KEYS: [&str; 3] = ["PARQUET:field_doc", "field_doc", "doc"];
+
+fn get_field_id(field: &Field) -> Option<i32> {
+    for key in ARROW_FIELD_ID_KEYS {
+        if let Some(value) = field.metadata().get(key) {
+            return value.parse::<i32>().ok();
+        }
+    }
+    None
+}
+
+fn get_field_doc(field: &Field) -> Option<String> {
+    for key in ARROW_FIELD_DOC_KEYS {
+        if let Some(value) = field.metadata().get(key) {
+            return Some(value.clone());
+        }
+    }
+    None
+}
+
+struct ArrowSchemaConverter {}
+
+impl ArrowSchemaConverter {
+    #[allow(dead_code)]
+    fn new() -> Self {
+        Self {}
+    }
+
+    fn convert_fields(fields: &Fields, field_results: &[Type]) -> Result<Vec<NestedFieldRef>> {
+        let mut results = Vec::with_capacity(fields.len());
+        for i in 0..fields.len() {
+            let field = &fields[i];
+            let field_type = &field_results[i];
+            let id = get_field_id(field).ok_or(Error::new(
+                ErrorKind::DataInvalid,
+                "Field id not found in metadata",
+            ))?;
+            let doc = get_field_doc(field);
+            let nested_field = NestedField {
+                id,
+                doc,
+                name: field.name().clone(),
+                required: !field.is_nullable(),
+                field_type: Box::new(field_type.clone()),
+                initial_default: None,
+                write_default: None,
+            };
+            results.push(Arc::new(nested_field));
+        }
+        Ok(results)
+    }
+}
+
+impl ArrowSchemaVisitor for ArrowSchemaConverter {
+    type T = Type;
+    type U = Schema;
+
+    fn schema(&mut self, schema: &ArrowSchema, values: Vec<Self::T>) -> Result<Self::U> {
+        let fields = Self::convert_fields(schema.fields(), &values)?;
+        let builder = Schema::builder().with_fields(fields);
+        builder.build()
+    }
+
+    fn r#struct(&mut self, fields: &Fields, results: Vec<Self::T>) -> Result<Self::T> {
+        let fields = Self::convert_fields(fields, &results)?;
+        Ok(Type::Struct(StructType::new(fields)))
+    }
+
+    fn list(&mut self, list: &DataType, value: Self::T) -> Result<Self::T> {
+        match list {
+            DataType::List(element_field) => {
+                let id = get_field_id(element_field).ok_or(Error::new(
+                    ErrorKind::DataInvalid,
+                    "Field id not found in metadata",
+                ))?;
+                let doc = get_field_doc(element_field);
+                let element_field = Arc::new(NestedField {
+                    id,
+                    doc,
+                    name: element_field.name().clone(),
```

Comment:
This is incorrect, list element field should always have field name "element".

Reply:
Good catch.